velox/core/PlanNode.h

```diff
@@ -522,13 +522,54 @@ class AggregationNode : public PlanNode {
     static Aggregate deserialize(const folly::dynamic& obj, void* context);
   };
 
+  // These fields are set if the aggregation has a GroupId input.
+  struct GroupingSetDescriptor {
+    // Output channel for the group_id column. The group_id column is
+    // populated with the globalGroupingSet number for global aggregation.
+    std::optional<column_index_t> groupIdChannel;
+
+    // These are the global grouping set numbers. Global grouping sets
+    // provide a global aggregation over all the grouping keys.
+    std::vector<vector_size_t> globalGroupingSets;
+
+    folly::dynamic serialize() const;
+
+    static GroupingSetDescriptor deserialize(
+        const folly::dynamic& obj,
+        void* context);
+  };
+
+  AggregationNode(
+      const PlanNodeId& id,
+      Step step,
+      const std::vector<FieldAccessTypedExprPtr>& groupingKeys,
+      const std::vector<FieldAccessTypedExprPtr>& preGroupedKeys,
+      const std::vector<std::string>& aggregateNames,
+      const std::vector<Aggregate>& aggregates,
+      bool ignoreNullKeys,
+      PlanNodePtr source);
+
   AggregationNode(
       const PlanNodeId& id,
       Step step,
       const std::vector<FieldAccessTypedExprPtr>& groupingKeys,
       const std::vector<FieldAccessTypedExprPtr>& preGroupedKeys,
       const std::vector<std::string>& aggregateNames,
       const std::vector<Aggregate>& aggregates,
+      const std::vector<vector_size_t>& globalGroupingSets,
+      const std::optional<const std::string>& groupIdName,
+      bool ignoreNullKeys,
+      PlanNodePtr source);
+
+  AggregationNode(
+      const PlanNodeId& id,
+      Step step,
+      const std::vector<FieldAccessTypedExprPtr>& groupingKeys,
+      const std::vector<FieldAccessTypedExprPtr>& preGroupedKeys,
+      const std::vector<std::string>& aggregateNames,
+      const std::vector<Aggregate>& aggregates,
+      const std::vector<vector_size_t>& globalGroupingSets,
+      const std::optional<column_index_t>& groupIdChannel,
```

Comment:
Can we add only one extra constructor taking either groupIdChannel or groupIdName? 

Is group-id an input column? If so, consider specifying it using FieldAccessTypedExprPtr that can be null.

Reply:
GroupIdChannel is an output column.

From the Prestissimo side, we want to send the groupIdVariable name from the protocol so I added that constructor. However, the ::create() function in AggregationNode can't use that constructor as it requires to get the name of the groupId column from the outputType which it doesn't create until after the node is constructed.

An alternative for me is to keep the groupIdVariable in the GroupingSetDescriptor instead of the channel. I'll update the code for this.

Reply:
@aditi-pandit 

AggregationNode is defined by specifying grouping keys + a set of aggregates. The output contains grouping keys followed by aggregations.

In a special case where Aggregation follows GroupId, one of the grouping keys is group-id column generated by the GroupId node. Hence, group-id is an input column for Aggregation and should be specified as FieldAccessTypedExprPtr.

The reason for specifying group-id column to Aggregation is to request it to produce 1 or more extra rows with global empty aggregation values in case there are no input rows. In this case, we specify a list of globalGroupingSets as group ID values and a group-id input column and expect Aggregation to produce one row per global grouping set with nulls in all grouping key columns except for group-id and empty global aggregation values. The values in globalGroupingSets list should be used to populate group-id column.

For example, given

* groupingKeys = [a, b, c, gid]
* globalGroupingSets = [1, 2, 3]
* aggregates = [sum(x), count(y)]
* groupId = gid

and empty input (zero rows), the aggregation result is expected to be

```
NULL, NULL , NULL, 1, NULL, 0
NULL, NULL , NULL, 2, NULL, 0
NULL, NULL , NULL, 3, NULL, 0
```

Reply:
@mbasmanova : Yes, this is the understanding for this change.

At the aggregation operator side, its important to know the output channel at which the groupId column is populated for this logic. Given the groupId column name (which is the same as its input name in the grouping keys), it can be used to lookup this output channel.

I have now used the groupId column name (instead of channel id) in the node so that some logic around groupIdChannel is simplified. The old logic was relying on knowing the outputType of the aggregationNode.

I feel adding a FieldAccessTypedExprPtr for this is un-necessary as we only need the column name.

Hope the current code is more satisfactory.